File: PokePlannerWeb/ClientApp/src/models/MoveLearnMethodEntry.ts
import { LocalString } from "./LocalString"

/**
 * Represents a move learn method in the data store.
 */
export class MoveLearnMethodEntry {
    /**
     * The ID of the move learn method.
     */
    moveLearnMethodId: number

    /**
     * The name of the move learn method.
     */
    name: string

    /**
     * The display names of the move learn method.
     */
    displayNames: LocalString[]

    /**
     * The descriptions of the move learn method.
     */
    descriptions: LocalString[]

    /**
     * Constructor.
     */
    constructor(
        moveLearnMethodId: number,
        name: string,
        displayNames: LocalString[],
        descriptions: LocalString[]
    ) {
        this.moveLearnMethodId = moveLearnMethodId
        this.name = name
        this.displayNames = displayNames
        this.descriptions = descriptions
    }

    /**
     * Returns a move learn method created from the given entry.
     */
    static from(method: MoveLearnMethodEntry) {
        return new MoveLearnMethodEntry(
            method.moveLearnMethodId,
            method.name,
            method.displayNames,
            method.descriptions
        )
    }

    /**
     * Returns the move learn method's display name in the given locale.
     */
    getDisplayName(locale: string): string | undefined {
        let localName = this.displayNames.find(n => n.language === locale)
        if (localName === undefined) {
            console.warn(
                `Move learn method ${this.moveLearnMethodId} is missing display name in locale '${locale}'`
            )
        }

        return localName?.value
    }

    /**
     * Returns the move learn method's description in the given locale.
     */
    getDescription(locale: string): string | undefined {
        let localDesc = this.descriptions.find(n => n.language === locale)
        if (localDesc === undefined) {
            console.warn(
                `Move learn method ${this.moveLearnMethodId} is missing description in locale '${locale}'`
            )
        }

        return localDesc?.value
    }
}

File: PokePlannerWeb.Data/DataStore/Models/MoveLearnMethodEntry.cs
﻿using System.Collections.Generic;

namespace PokePlannerWeb.Data.DataStore.Models
{
    /// <summary>
    /// Represents a move learn method in the data store.
    /// </summary>
    public class MoveLearnMethodEntry : NamedApiResourceEntry
    {
        /// <summary>
        /// Gets the ID of the move learn method.
        /// </summary>
        public int MoveLearnMethodId => Key;

        /// <summary>
        /// Gets or sets the display names of the move learn method.
        /// </summary>
        public List<LocalString> DisplayNames { get; set; }

        /// <summary>
        /// Gets or sets the descriptions of the move learn method.
        /// </summary>
        public List<LocalString> Descriptions { get; set; }
    }
}
